### DISCUSSION TITLE
What is the philosophy of the stdlib?

### DISCUSSION THREAD

**Discussion by lsh** [modular/modular#92] [Q&A]
Each language has an opinion on how large the standard library should be, and what belongs in it. Go has a solid standard library (solid enough that Deno ported it for their use). Rust is an example of a relatively small standard library, where even random number generation and async runtimes are left up to the community.

What is Mojo's philosophy? Would something like a data frame make it into the stdlib? How about the ability to fetch data using HTTP requests?

As people begin to open tickets about modules they think belong in the standard library, it's probably good to have some guiding information about what makes more sense as community projects.

**Comment by timmermansjoy** [comment]
@lsh since mojo is going to be a superset of python you can assume the python stdlib will be integrated. But for the rest there is some wiggle room on what to implement. But I would say that there will come some community packages for dataframes so the core team can focus on the language and the community to build with the language

**Comment by Mogball** [comment]
> What is the philosophy of the stdlib?

Short answer: there isn't one, yet. You are right that we should develop one, however.

**Reply by lsh** [reply]
[Related discussion from the Python Language Summit 2023](https://pyfound.blogspot.com/2023/05/the-python-language-summit-2023-what-is.html?m=1), which may help to shape Mojo's decisions.

**Reply by lattner** [reply]
nice article!

**Comment by ksandvik** [comment]
Could this stdlib have defined/standard traits so it's easy to extend it or build upon it (similar to stdlib in Rust and crates)?

**Reply by lattner** [reply]
Yes absolutely, that's the goal, but we need to implement support for traits first :)

**Comment by ksandvik** [comment]
With a default way to define package dependencies (i.e. such as  pyproject.toml) and a good packaging tool, and traits, then multiple std implementations could co-exist as long as it's easy to packaging this into a single executable or executable environment.

**Reply by epogrebnyak** [reply]
Multiple std implementations do you mean like a prelude? Some module that is not stdlib bit loaded by default?

**Comment by ksandvik** [comment]
With traits you could write the same library components as two different implementations, see Rust Tokyo vs Async-std as one example for std async libraries.

**Comment by helehex** [comment]
Maybe if it satisfies one of these:

1. Python people expect to have it
2. It's expected to be used in more than half of all projects
3. There's typically a best practice implementation, and its used in roughly a quarter of projects
